File: chess_in_programming/1st-edition/chess queen and king/sol.py
def beats(wk, wq, bk):
  xq = ord(wq[0]) - ord('a') + 1
  yq = int(wq[1])
  xwk = ord(wk[0]) - ord('a') + 1
  ywk = int(wk[1])
  xbk = ord(bk[0]) - ord('a') + 1
  ybk = int(bk[1])

  if xq < 0 or xq > 8 or yq < 0 or yq > 8 or xwk < 0 or xwk > 8 or ywk < 0 or ywk > 8 or xbk < 0 or xbk > 8 or ybk < 0 or ybk > 8:
    return False

  if xq == xbk and (xwk != xbk or not (yq <= ywk and ywk <= ybk or yq >= ywk and ywk >= ybk)) or yq == ybk and (ywk != ybk or not (xq <= xwk and xwk <= xbk or xq >= xwk and xwk >= xbk)):
    return True

  if abs(xq - xbk) == abs(yq - ybk):
    tx = xq
    ty = yq
    if xq < xbk and yq < ybk:
      for i in range(1, abs(xq - xbk)):
        if tx + i == xwk and ty + i == ywk:
          return False
    elif xq < xbk and yq > ybk:
      for i in range(1, abs(xq - xbk)):
        if tx + i == xwk and ty - i == ywk:
          return False
    elif xq > xbk and yq < ybk:
      for i in range(1, abs(xq - xbk)):
        if tx - i == xwk and ty + i == ywk:
          return False
    else:
      for i in range(1, abs(xq - xbk)):
        if tx - i == xwk and ty - i == ywk:
          return False

    return True

  return False


def main():
  s = input()
  if len(s) != 8:
    print("NO")
    return

  if beats(s[0:2], s[2:4], s[4:6]):
    print("YES")
  else:
    print("NO")


if __name__ == "__main__":
  main()

File: chess_in_programming/1st-edition/chess queen and king/sol.java
public static boolean beats(String wk, String wq, String bk) {
  int xq = wq.charAt(1) - '0', yq = wq.charAt(0) - 'a' + 1, xwk = wk.charAt(1) - '0', ywk = wk.charAt(0) - 'a' + 1, xbk = bk.charAt(1) - '0', ybk = bk.charAt(0) - 'a' + 1;
  if (xq < 0 || xq > 8 || yq < 0 || yq > 8 || xwk < 0 || xwk > 8 || ywk < 0 || ywk > 8 || xbk < 0 || xbk > 8 || ybk < 0 || ybk > 8) {
    return false;
  }
  if ((xq == xbk && (xwk != xbk || !(yq <= ywk && ywk <= ybk || yq >= ywk && ywk >= ybk)) || (yq == ybk && (ywk != ybk || !(xq <= xwk && xwk <= xbk || xq >= xwk && xwk >= xbk))))) {
    return true;
  }
  if (Math.abs(xq - xbk) == Math.abs(yq - ybk)) {
    int tx = xq, ty = yq;
    if (xq < xbk && yq < ybk) {
      for (int i = 1; i < Math.abs(xq - xbk); i++) {
        if (tx + i == xwk && ty + i == ywk) {
          return false;
        }
      }
    } else if (xq < xbk && yq > ybk) {
      for (int i = 1; i < Math.abs(xq - xbk); i++) {
        if (tx + i == xwk && ty - i == ywk) {
          return false;
        }
      }
    } else if (xq > xbk && yq < ybk) {
      for (int i = 1; i < Math.abs(xq - xbk); i++) {
        if (tx - i == xwk && ty + i == ywk) {
          return false;
        }
      }
    } else {
      for (int i = 1; i < Math.abs(xq - xbk); i++) {
        if (tx - i == xwk && ty - i == ywk) {
          return false;
        }
      }
    }
    return true;
  }
  return false;
}

public static void main(String[] args) {
  String s = "";
  try (Scanner sc = new Scanner(System.in)) {
    s = sc.nextLine();
  }
  if (s.length() != 8) {
    System.out.println("NO");
    return;
  }
  if (beats(s.substring(0, 2), s.substring(3, 5), s.substring(6, 8))) {
    System.out.println("YES");
  } else {
    System.out.println("NO");
  }
}
